client/service/src/task_manager.rs

```diff
@@ -0,0 +1,148 @@
+}
+
+impl SpawnTaskHandle {
+	/// Spawns the given task with the given name.
+	pub fn spawn(&self, name: impl Into<Cow<'static, str>>, task: impl Future<Output = ()> + Send + 'static) {
+		let on_exit = self.on_exit.clone();
+		let future = async move {
+			futures::pin_mut!(task);
+			let _ = select(on_exit, task).await;
+		};
+		if self.sender.unbounded_send((Box::pin(future), name.into())).is_err() {
+			error!("Failed to send task to spawn over channel");
+		}
+	}
+}
+
+impl Spawn for SpawnTaskHandle {
+	fn spawn_obj(&self, future: FutureObj<'static, ()>)
+	-> Result<(), SpawnError> {
+		let future = select(self.on_exit.clone(), future).map(drop);
+		self.sender.unbounded_send((Box::pin(future), From::from("unnamed")))
+			.map_err(|_| SpawnError::shutdown())
+	}
+}
+
+type Boxed01Future01 = Box<dyn futures01::Future<Item = (), Error = ()> + Send + 'static>;
+
+impl futures01::future::Executor<Boxed01Future01> for SpawnTaskHandle {
+	fn execute(&self, future: Boxed01Future01) -> Result<(), futures01::future::ExecuteError<Boxed01Future01>>{
+		self.spawn("unnamed", future.compat().map(drop));
+		Ok(())
+	}
+}
+
+/// Helper struct to manage background/async tasks in Service.
+pub struct TaskManager {
+	/// A future that resolves when the service has exited, this is useful to
+	/// make sure any internally spawned futures stop when the service does.
+	on_exit: exit_future::Exit,
+	/// A signal that makes the exit future above resolve, fired on service drop.
+	signal: Option<Signal>,
+	/// Sender for futures that must be spawned as background tasks.
+	to_spawn_tx: TaskScheduler,
+	/// Receiver for futures that must be spawned as background tasks.
+	to_spawn_rx: mpsc::UnboundedReceiver<(Pin<Box<dyn Future<Output = ()> + Send>>, Cow<'static, str>)>,
+}
+
+impl TaskManager {
+	/// New task manager for service.
+	pub fn new() -> Self {
+		let (signal, on_exit) = exit_future::signal();
+		let (to_spawn_tx, to_spawn_rx) = mpsc::unbounded();
+		Self {
+			on_exit,
+			signal: Some(signal),
+			to_spawn_tx,
+			to_spawn_rx,
+		}
+	}
+
+	pub(crate) fn spawn_handle(&self) -> SpawnTaskHandle {
+		SpawnTaskHandle {
+			on_exit: self.on_exit.clone(),
+			sender: self.to_spawn_tx.clone(),
+		}
+	}
+
+	/// Spawn background/async task, which will be aware on exit signal.
+	pub fn spawn(&self, name: impl Into<Cow<'static, str>>, task: impl Future<Output = ()> + Send + 'static) {
+		let on_exit = self.on_exit.clone();
+		let future = async move {
+			futures::pin_mut!(task);
+			let _ = select(on_exit, task).await;
+		};
+		if self.to_spawn_tx.unbounded_send((Box::pin(future), name.into())).is_err() {
+			error!("Failed to send task to spawn over channel");
+		}
+	}
+
+	/// Get sender where background/async tasks can be sent.
+	pub fn scheduler(&self) -> TaskScheduler {
+		self.to_spawn_tx.clone()
+	}
+
+	/// Process background task receiver
+	pub(crate) fn process_receiver(&mut self, executor: &ServiceTaskExecutor, cx: &mut Context) {
```

Comment:
This is also an implementation detail and generally a bad API. I'm really not sure it's a good idea to extract this struct in a separate module.

Reply:
Totally agree,

Unfortunately, this is here because of the way actual ServiceTaskExecutor is instantiated (in `run_until_exit`. Otherwise TaskManager could've own executor and never expose it.

Reply:
Okay, I decoupled this

But it had required some more code.. :)